packages/@lwc/engine-dom/src/customElements.ts

```diff
@@ -0,0 +1,128 @@
+export interface UpgradableCustomElementConstructor extends CustomElementConstructor {
+    new (upgradeCallback: UpgradeCallback): HTMLElement;
+}
+
+let HTMLElementConstructor;
+let getUpgradableElement: (tagName: string) => CustomElementConstructor;
+let getUserConstructor: (
+    upgradeCallback: UpgradeCallback,
+    connectedCallback?: UpgradeCallback,
+    disconnectedCallback?: UpgradeCallback
+) => UpgradableCustomElementConstructor | UpgradeCallback;
+
+function isCustomElementRegistryAvailable() {
+    if (typeof customElements === 'undefined') {
+        return false;
+    }
+    try {
+        // dereference HTMLElement global because babel wraps globals in compat mode with a
+        // _wrapNativeSuper()
+        // This is a problem because LWCUpgradableElement extends renderer.HTMLElement which does not
+        // get wrapped by babel.
+        const HTMLElementAlias = HTMLElement;
+        // In case we use compat mode with a modern browser, the compat mode transformation
+        // invokes the DOM api with an .apply() or .call() to initialize any DOM api sub-classing,
+        // which are not equipped to be initialized that way.
+        class clazz extends HTMLElementAlias {}
+        customElements.define('lwc-test-' + Math.floor(Math.random() * 1000000), clazz);
+        new clazz();
+        return true;
+    } catch {
+        return false;
+    }
+}
+
+if (isCustomElementRegistryAvailable()) {
+    const defineScopedElement = createScopedRegistry();
+
+    // It's important to cache window.HTMLElement immediately after calling createScopedRegistry().
+    // Otherwise, someone else could overwrite window.HTMLElement (e.g. another copy of the engine, or another pivot
+    // implementation) and we would get "Illegal constructor" errors because the HTMLElement prototypes are mixed up.
+    const { HTMLElement } = window;
+
+    HTMLElementConstructor = HTMLElement;
+    const cachedConstructor: Record<string, CustomElementConstructor> = create(null);
+
+    getUpgradableElement = (tagName: string) => {
+        let Ctor = cachedConstructor[tagName];
+        if (!Ctor) {
+            // TODO [#2972]: this class should expose observedAttributes as necessary
+            class LWCUpgradableElement extends HTMLElement {}
+            Ctor = defineScopedElement(tagName, LWCUpgradableElement);
+        }
+        return Ctor;
+    };
+    getUserConstructor = (upgradeCallback: UpgradeCallback) => {
+        class UserElement extends HTMLElement {
+            constructor() {
+                super();
+                upgradeCallback(this);
+            }
+        }
+        if (features.ENABLE_NATIVE_CUSTOM_ELEMENT_LIFECYCLE) {
+            (UserElement.prototype as any).connectedCallback = function () {
+                connectRootElement(this);
+            };
+            (UserElement.prototype as any).disconnectedCallback = function () {
+                disconnectRootElement(this);
+            };
+        }
```

Comment:
Native custom element lifecycle only makes sense 1) in the browser and 2) when native custom elements are available. So it seems best to put this here.